job-ingest: add new plugin-based job validator

grondo opened the pull request:
This PR adds a new `flux job-validator` command which replaces the previous approach of loading different validator scripts at runtime via the job-ingest `validator=PATH` option. Instead `flux job-validator` supports loading one or more validator plugins at runtime, and runs each selected plugin simultaneously in a thread pool. If any one validation routine fails, then the job is rejected.

To allow the most flexibility in developing validation plugins, instead of sending only the submitted jobspec to `flux job-validator`, the job-ingest module now includes all other parameters of submission, including the submitting userid, rolemask, flags, and urgency.

Validation plugins are discovered by default via a [namespace package](https://packaging.python.org/guides/creating-and-discovering-plugins/#using-namespace-packages): `flux.job.validator.plugins`. Currently, the old `jobspec` and `schema` validators are packaged by default in this namespace. If another framework project wants to add a new validator, all it needs to do is to install `<name>.py` somewhere in `PYTHONPATH` where it will discovered in the namspace. The new plugin can then be activated via a new `validator-plugins=LIST` option in the job-ingest module. (TODO: allow validator plugins to be configured via TOML config as well). Note, `flux job-validator` also supports loading plugins from a specific file, or extending `sys.path` only when loading plugins, both of which are useful for testing or when adding plugins to the default PYTHONPATH is impractical.

Plugins are still configured via argparse command-line options. `flux job-validator` parses the command-line in two phases -- the first phase parses a `--plugins` option which tells the validator which plugins to activate. When plugins are activated, their `configure` method is called and they can add more options to the command-line. After all selected plugins are active, then `argparse.parse_args()` is called again, allowing plugin-specific cmdline opts to be parsed. This also allows `--plugins=LIST --help` to work as expected -- the plugin options are included in the help output, e.g.:

```
$ flux job-validator --plugins=schema,jobspec --help
usage: flux-job-validator [--jobspec-only] [--list-plugins]
                          [--plugins PLUGINS] [-h] --schema SCHEMA
                          [--require-version V]

Validate Flux jobs from lines of JSON input on stdin

optional arguments:
  -h, --help

Program options:
      --jobspec-only       Expect only JSON encoded jobspec on stdin
      --list-plugins       List available validator plugins and exit

Validator options:
      --plugins=PLUGINS

Options provided by plugins:
      --schema=SCHEMA      Validate with SCHEMA
      --require-version=V  Require jobspec version V
```

Validator plugins may perform arbitrary work when validating jobs. For example, one item todo here will be to add a validator plugin which contacts the loaded scheduler to check if the job is feasible. There is a mechanism in place for each thread to get an on-demand Flux handle which can be used to execute RPCs if necessary.

Since loading arbitrary validator scripts is no longer necessary, the `validator=` option in job-ingest is dropped after updating the testsuite accordingly.

This PR is not quite done, as it needs a little testing and the addition of perhaps a couple other validator plugins as examples. However, I wanted to post it early to get feedback on the general approach.

grondo commented:
This PR has been updated with a demo job validation plugin which supports feasibility checks at ingest.

A new `feasibility` validator was added here which, by default, sends job information to an optional `sched.feasibility` service, giving the loaded scheduler the ability to reject jobs during validation. A simple handler for `sched.feasibility` is added to sched-simple to demonstrate. Jobs are now rejected at ingest if they are not feasible, or are otherwise invalid for sched-simple, when the `feasibility` plugin is activated, e.g.

```
ƒ(s=1,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini run -n100 hostname
0.014s: job.exception type=alloc severity=0 unsatisfiable request
ƒ(s=1,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini run -g 1 hostname
0.014s: job.exception type=alloc severity=0 Unsupported resource type 'gpu'

ƒ(s=1,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux module reload job-ingest validator-plugins=jobspec,feasibility

ƒ(s=1,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini run -n100 hostname 
flux-mini: ERROR: [Errno 75] request is not satisfiable
ƒ(s=1,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini run -g 1 hostname
flux-mini: ERROR: [Errno 22] Unsupported resource type 'gpu'
```

To demonstrate the simplicity of a job validator plugin, here's the working `feasibility` plugin. Note this plugin would be much smaller without addition of the `--feasibility-service=NAME` option

```python
##############################################################
# Copyright 2021 Lawrence Livermore National Security, LLC
# (c.f. AUTHORS, NOTICE.LLNS, COPYING)
#
# This file is part of the Flux resource manager framework.
# For details, see https://github.com/flux-framework.
#
# SPDX-License-Identifier: LGPL-3.0
##############################################################

"""Use sched.feasibility RPC to validate job

This plugin validates jobs via a sched.feasibility RPC.

This allows jobs which are making infeasible or otherwise invalid
requests to be rejected by the scheduler before ingest, instead of
when the scheduler attempts to allocate resources for them.

The RPC endpoint for feasibility checks can be set with an optional
--feasibility-service=NAME option.

ENOSYS errors from the RPC are ignored, in case the loaded scheduler
does not support the sched.feasibility RPC
"""

import errno
from flux.job.validator import ValidatorPlugin


class Validator(ValidatorPlugin):
    def __init__(self, parser):
        self.service_name = "sched.feasibility"
        parser.add_argument(
            "--feasibility-service",
            metavar="NAME",
            help="Set feasibility RPC service endpoint "
            f"(default={self.service_name})",
        )

    def configure(self, args):
        if args.feasibility_service:
            self.service_name = args.feasibility_service

    def validate(self, args):
        try:
            args.flux.rpc(self.service_name, args.jobinfo).get()
        except OSError as err:
            if err.errno == errno.ENOSYS:
                #  Treat ENOSYS as success
                return (0, None)
            return (err.errno, err.strerror)

```

chu11 commented on t/t2200-job-ingest.t:
```diff
@@ -21,10 +21,7 @@ Y2J="flux python ${JOBSPEC}/y2j.py"
-JSONSCHEMA_VALIDATOR=${FLUX_SOURCE_DIR}/src/modules/job-ingest/validators/validate-schema.py
-FAKE_VALIDATOR=${SHARNESS_TEST_SRCDIR}/ingest/fake-validate.sh
-BAD_VALIDATOR=${SHARNESS_TEST_SRCDIR}/ingest/bad-validate.sh
+BAD_VALIDATOR=${SHARNESS_TEST_SRCDIR}/ingest/bad-validate.py
```
forgot to add bad-validate.py to commit?

grondo replied:
Ooh, yeah good catch. And since the test expects failure, the test still "passes". That's not so good.

chu11 reviewed (commented):
Took a quick skim, mostly trying to understand how this all works.  Overall, I like it.  One comment:

```
$ flux job-validator --plugins=schema,jobspec --help
usage: flux-job-validator [--jobspec-only] [--list-plugins]
                          [--plugins PLUGINS] [-h] --schema SCHEMA
                          [--require-version V]
```

do we need a way to see what is currently configured for job validation?

grondo commented:
> do we need a way to see what is currently configured for job validation? Possibly with options/args users picked?

Yeah, that is a good idea and I had considered adding that at the same time as adding support for toml configuration. 

At this point I'm wondering if we should merge this PR once I've expanded the testing a bit, and add configuration support later. We didn't have support for querying the validator executable nor args before, so nothing is lost here, and we might want time to decide how to structure the configuration or if tuning of validators should somehow be allowed from the cmdline when launching a sub-instance, etc.

garlick commented on src/cmd/flux-job-validator.py:
```diff
@@ -0,0 +1,109 @@
+    )
+    script_group = parser.add_argument_group("Program options")
+    script_group.add_argument(
+        "--jobspec-only",
```
Neat!  Really neat.

I wonder if it would be interesting to users to be able to play with feasibility without actually submitting anything?   Maybe we could add a `--feasibility` option like `--dry-run` to `flux mini submit` that does the equivalent of
```
$ flux mini submit --dry-run hostname | flux job-validator --jobspec-only --plugins feasibility
{"errnum": 0}
```
and parses the returned JSON?   Just a thought (but probably not for this PR)

grondo replied:
That's a good idea! The validator is actually a python class now so it could embedded directly in tools like `flux-mini.py` (which was part of the reason for doing it that way)

garlick reviewed (approved):
Cool!  Just a couple of spelling errors in my review.  This seems like a really nice approach to me, albeit my python is sub-par, so some of the review was just sightseeing, looking at all the pretty things.

Extra points awarded for closing a 6 year old bug!

grondo commented:
Ok, I've pushed a few updates in the testsuite:

 - Split `t2200-job-ingest.t` into two tests (and renamed them to `t21*` so they can be numerically adjacent)
 - `t2100-job-ingest.t` contains the job-ingest specific tests and use the dummy job-manger module, and the validator tests were moved to `t2110-job-ingest-validator.t` and use the "job" rc personality so that the feasibility validator can be used.
 - The `test_valid()` and `test_invalid()` functions were updated to run the tests as a series of Flux jobs instead of serially converting yaml and attempting submission (unfortunately only a small gain here)

I opened two issues on the remaining work to be done #3542 (allow query of validator command and args) and #3541 (allow validator toml configuration)

garlick commented:
Looks great @grondo.  Please set MWP when you're ready.

grondo commented:
Thanks!
